presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToJsonCast.java

```diff
@@ -66,23 +72,29 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         checkArgument(arity == 1, "Expected arity to be 1");
         Type type = boundVariables.getTypeVariable("T");
         Type arrayType = typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(TypeSignatureParameter.of(type.getTypeSignature())));
+        checkCondition(canCastToJson(arrayType), INVALID_CAST_ARGUMENT, "Cannot cast %s to JSON", arrayType);
 
-        MethodHandle methodHandle = METHOD_HANDLE.bindTo(arrayType);
+        JsonUtil.JsonGeneratorWriter writer = JsonUtil.JsonGeneratorWriter.createJsonGeneratorWriter(type);
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(writer);
         return new ScalarFunctionImplementation(false, ImmutableList.of(false), methodHandle, isDeterministic());
     }
 
-    public static Slice toJson(Type arrayType, ConnectorSession session, Block array)
+    public static Slice toJson(JsonUtil.JsonGeneratorWriter writer, ConnectorSession session, Block block)
     {
-        Type elementType = arrayType.getTypeParameters().get(0);
-        List<Object> objectValues = new ArrayList<>(array.getPositionCount());
-        for (int i = 0; i < array.getPositionCount(); i++) {
-            objectValues.add(JsonFunctions.getJsonObjectValue(elementType, session, array, i));
-        }
         try {
-            return Slices.utf8Slice(OBJECT_MAPPER.get().writeValueAsString(objectValues));
+            SliceOutput output = new DynamicSliceOutput(100);
+            try (JsonGenerator jsonGenerator = createJsonGenerator(JSON_FACTORY, output)) {
+                jsonGenerator.writeStartArray();
+                for (int i = 0; i < block.getPositionCount(); i++) {
+                    writer.writeJsonValue(jsonGenerator, block, i, session);
+                }
+                jsonGenerator.writeEndArray();
```

Comment:
These 5 lines are pretty much inlining `JsonUtil.writeJsonValue`. What do you think about delegating to it instead?